File: 0472-concatenated-words/0472-concatenated-words.py
class Solution:
    def dfs(self, word: str, length: int, visited: List[bool], dictionary: Set[str]) -> bool:
        if length == len(word):
            return True
        if visited[length]:
            return False
        visited[length] = True
        start = 1 if length == 0 else 0
        for i in range(len(word) - start, length, -1):
            if word[length:i] in dictionary and self.dfs(word, i, visited, dictionary):
                return True
        return False
    
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        
        dictionary = {word for word in words}
        ans = []
        for word in words:
            visited = [False] * len(word)
            if (self.dfs(word, 0, visited, dictionary)):
                ans.append(word)
        return ans

File: 0472-concatenated-words/0472-concatenated-words.cpp
class Solution {
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        // DP with hashmap
        set<string> dictionary(words.begin(), words.end());
        vector<string> ans;
        
        for (const string& word: words){
            const int n = word.size();
            vector<bool> dp(n+1);
            dp[0] = true;
            
            for (int i = 1; i <= n; ++i){
                for (int j = (i == n? 1 : 0); !dp[i] && j < i; ++j){
                    dp[i] = dp[j] && dictionary.count(word.substr(j, i - j));
                }
            }
            
            if (dp[n]) ans.push_back(word);
        }
        
        return ans;
    }
};